File: backend/app/routers/keywords.py
import uuid
from datetime import datetime, timezone

from fastapi import APIRouter, HTTPException, status

from app.mock_store import keywords
from app.schemas.keyword import KeywordCreate, KeywordRead, KeywordUpdate

router = APIRouter(prefix="/keywords", tags=["keywords"])


@router.get("/", response_model=list[KeywordRead])
async def list_keywords():
    return sorted(keywords.values(), key=lambda x: x["created_at"], reverse=True)


@router.post("/", response_model=KeywordRead, status_code=status.HTTP_201_CREATED)
async def create_keyword(body: KeywordCreate):
    kw = {
        "id": uuid.uuid4(),
        "phrase": body.phrase.strip().lower(),
        "is_active": body.is_active,
        "created_at": datetime.now(timezone.utc),
    }
    keywords[kw["id"]] = kw
    return kw


@router.get("/{keyword_id}", response_model=KeywordRead)
async def get_keyword(keyword_id: uuid.UUID):
    kw = keywords.get(keyword_id)
    if not kw:
        raise HTTPException(status_code=404, detail="Keyword not found")
    return kw


@router.put("/{keyword_id}", response_model=KeywordRead)
async def update_keyword(keyword_id: uuid.UUID, body: KeywordUpdate):
    kw = keywords.get(keyword_id)
    if not kw:
        raise HTTPException(status_code=404, detail="Keyword not found")
    data = body.model_dump(exclude_unset=True)
    if "phrase" in data:
        data["phrase"] = data["phrase"].strip().lower()
    kw.update(data)
    return kw


@router.delete("/{keyword_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_keyword(keyword_id: uuid.UUID):
    if keyword_id not in keywords:
        raise HTTPException(status_code=404, detail="Keyword not found")
    del keywords[keyword_id]

File: backend/src/routers/keywords.js
'use strict';

const { Router } = require('express');
const { v4: uuidv4 } = require('uuid');
const { keywords } = require('../mockStore');

const router = Router();

// GET /api/keywords
router.get('/', (_req, res) => {
  const items = [...keywords.values()].sort((a, b) => b.created_at.localeCompare(a.created_at));
  res.json(items);
});

// POST /api/keywords
router.post('/', (req, res) => {
  const { phrase, is_active = true } = req.body;
  if (!phrase || !phrase.trim()) {
    return res.status(422).json({ detail: 'phrase is required' });
  }
  const kw = { id: uuidv4(), phrase: phrase.trim().toLowerCase(), is_active, created_at: new Date().toISOString() };
  keywords.set(kw.id, kw);
  res.status(201).json(kw);
});

// GET /api/keywords/:id
router.get('/:id', (req, res) => {
  const kw = keywords.get(req.params.id);
  if (!kw) return res.status(404).json({ detail: 'Keyword not found' });
  res.json(kw);
});

// PUT /api/keywords/:id
router.put('/:id', (req, res) => {
  const kw = keywords.get(req.params.id);
  if (!kw) return res.status(404).json({ detail: 'Keyword not found' });
  if (req.body.phrase !== undefined) kw.phrase = req.body.phrase.trim().toLowerCase();
  if (req.body.is_active !== undefined) kw.is_active = req.body.is_active;
  res.json(kw);
});

// DELETE /api/keywords/:id
router.delete('/:id', (req, res) => {
  if (!keywords.has(req.params.id)) return res.status(404).json({ detail: 'Keyword not found' });
  keywords.delete(req.params.id);
  res.status(204).end();
});

module.exports = router;
